File: 0946-validate-stack-sequences/0946-validate-stack-sequences.js
/**
 * @param {number[]} pushed
 * @param {number[]} popped
 * @return {boolean}
 */
var validateStackSequences = function(pushed, popped) {
    const n = pushed.length;
        const sk = new Stack();

        let l = 0;
        let idx = 0;

        while (l < n && idx < n) {

            sk.push(pushed[l++]);

            while (!sk.isEmpty() && sk.peek() == popped[idx]) {
                sk.pop();
                idx++;
            }
        }

        return sk.isEmpty();
};

File: 0946-validate-stack-sequences/0946-validate-stack-sequences.java
class Solution {
    public boolean validateStackSequences(int[] pushed, int[] popped) {
        Stack<Integer> sk = new Stack<>();
        int l = 0;
        int idx = 0;
        while (l < pushed.length && idx < popped.length) {
            sk.push(pushed[l++]);
            while (!sk.isEmpty() && sk.peek() == popped[idx]) {
                sk.pop();
                idx++;
            }
        }
        return sk.isEmpty();
    }
}